Production/src/Selectors.cc

```diff
@@ -177,5 +179,94 @@ TauJetSelector::Result TauJetTag::Select(const edm::Event& event, const std::deq
+        	tagObject->p4 = ref_muon->polarP4();
+       		tagObject->charge = ref_muon->charge();
+        	tagObject->id = unsigned(ref_muon->isLooseMuon()) * 1 + unsigned(ref_muon->isMediumMuon()) * 2
+                          	+ unsigned(ref_muon->isTightMuon(primaryVertex)) * 4;
+        	tagObject->isolation = PFRelIsolation(*ref_muon);
+        	tagObject->has_extramuon = hasExtraMuon(muons, ref_muon, primaryVertex);
+        	tagObject->has_extraelectron = hasExtraElectron(electrons, rho);
+        	tagObject->has_dimuon = hasExtraDimuon(muons, primaryVertex);
+    	}
+    	//	
+    	else if(ref_electron) { // Fill tag with ele if ele found rather than muon
+		tagObject->kind = tau_analysis::selectors::TagObject::Kind::Electron; //new tag object member
+        	tagObject->p4 = ref_electron->polarP4();
+        	tagObject->charge = ref_electron->charge();
+        	tagObject->id = ref_electron->electronID("mvaEleID-Fall17-noIso-V2-wp90"); //Not sure what to put here 
+        	tagObject->isolation = PFRelIsolation(*ref_electron,rho);
+        	tagObject->has_extramuon = hasExtraMuon(muons, ref_muon, primaryVertex);
+        	tagObject->has_extraelectron = hasExtraElectron(electrons, rho);
+        	tagObject->has_dimuon = hasExtraDimuon(muons, primaryVertex);	
+    	}
+    	else { // high MET case
+        	tagObject = nullptr;   
+    	}
+        const TauJet* selected = nullptr;
+        for(const TauJet& tauJet : tauJets) { //Look for highest pT jet
+            if(!tauJet.jet) continue;
+            const pat::Jet& jet = *tauJet.jet;
+            if(!(std::abs(jet.eta()) < 2.8) || ((ref_muon || ref_electron) && !(reco::deltaR(jet.p4(), tagObject->p4) > 0.8)))
+                    continue;
+            if(!selected || selected->jet->pt() < jet.pt()) {
+                    selected = &tauJet;
+                    selectedTauJets = { selected };
+            }
+        }
+    }
+    else { // No mu or ele tag nor high MET 
+        tagObject = nullptr;
+        for(const TauJet& tauJet : tauJets) {
+            if(!(tauJet.genLepton) || !(std::abs(tauJet.genLepton->visibleP4().eta()) < 2.8) || !(tauJet.genJet && std::abs(tauJet.genJet->eta()) < 2.8)) continue;
+	    selectedTauJets.push_back(&tauJet); //Fill with jets passing selection
+	}
+        std::sort(selectedTauJets.begin(), selectedTauJets.end(), [](auto &i, auto &j){return i->genLepton->visibleP4().pt() < j->genLepton->visibleP4().pt();});
+        selectedTauJets.resize(std::min(static_cast<int>(selectedTauJets.size()), 2));
+    }
+   
+    // Add true taus to selectedTauJets for all cases
+    for(const TauJet& tauJet : tauJets) {
+       if(!(tauJet.genLepton) || !(std::abs(tauJet.genLepton->visibleP4().eta()) < 2.8 && tauJet.genLepton->kind() == reco_tau::gen_truth::GenLepton::Kind::TauDecayedToHadrons)) continue;
+         selectedTauJets.push_back(&tauJet); //Fill with jets passing selection
+    }
+    //
+    return Result(selectedTauJets, tagObject);
```

Comment:
In the following part there will be no prioritization to the genLepton->kind (no `kind: tau_h > tau_e > tau_mu > e > mu > jet`) do we want to have it?

Also: the following part does require having >2 valid TauJets, from Konstantin's slides I had understanding it was needed.